pkg/alertmanager/alertstore/bucketclient/bucket_client.go

```diff
@@ -212,3 +286,7 @@ func (s *BucketAlertStore) getUserBucket(userID string) objstore.Bucket {
 func (s *BucketAlertStore) getAlertmanagerUserBucket(userID string) objstore.Bucket {
 	return bucket.NewUserBucketClient(userID, s.amBucket, s.cfgProvider).WithExpectedErrs(s.amBucket.IsObjNotFoundErr)
 }
+
+func (s *BucketAlertStore) getUserGrafanaBucket(userID string) objstore.Bucket {
+	return bucket.NewSSEBucketClient(userID, s.grafanaAMBucket, s.cfgProvider).WithExpectedErrs(s.amBucket.IsObjNotFoundErr)
```

Comment:
`NewSSEBucketClient` doesn't put the tenant ID in the key of the object (`NewUserBucketClient` used above does that). This comment claims that the key includes the user ID. Can you add an explanation for why we're not using a UserBucket?

Reply:
I'm happy to add the reasoning but I don't see the comment you're referring to (if I need to modify it) - am I missing something?

Reply:
ah, sorry. I forgot to paste the link. I meant this comment https://github.com/grafana/mimir/blob/083abea0147e15612272e7bfab850550f649e3db/pkg/alertmanager/alertstore/bucketclient/bucket_client.go#L37-L40

Reply:
so with the current approach we'd store a single grafana config per mimir cluster

https://github.com/grafana/mimir/blob/083abea0147e15612272e7bfab850550f649e3db/pkg/alertmanager/alertstore/bucketclient/bucket_client.go#L158

this line produces the same object key for the AM config for two different tenants. So one would overwrite the config of the other. Was this the intention?

regarding the reason for using an SSEBucket:

https://github.com/grafana/mimir/blob/083abea0147e15612272e7bfab850550f649e3db/pkg/alertmanager/alertstore/bucketclient/bucket_client.go#L291

You can still store multiple items per user as long as the objects have different keys, and they already do - `grafana_config` and `grafana_fullstate`.

I have the feeling we're talking about different things here 😅

Reply:
Woah, you were absolutely right - this didn't do what I thought it did. I think that in my initial testing, I used `NewUserBucketClient` and changed the approach eventually without actual validation.

Thank you so much for spotting this, it probably saved me hours of debugging and fixing. I've:

- Changed the approach to use `NewUserBucketClient`
- Add unit tests that validate the integrity of the storage key
- Modified the integration tests to cover for any sort of regression

All of it was done as part of https://github.com/grafana/mimir/pull/6682/commits/c919def819b5c4ea12e1821f390f35a506d348af